File: firebase_auth_oauth/android/src/main/kotlin/me/amryousef/apple/auth/firebase_auth_oauth/FirebaseAuthOAuthPluginError.kt
package me.amryousef.apple.auth.firebase_auth_oauth

import io.flutter.plugin.common.MethodChannel.Result

sealed class FirebaseAuthOAuthPluginError {
    data class FirebaseAuthError(val exception: Exception) : FirebaseAuthOAuthPluginError()
    data class PlatformError(val exception: Exception) : FirebaseAuthOAuthPluginError()
    data class PluginError(val error: String) : FirebaseAuthOAuthPluginError()

    val code: String
        get() {
            return when (this) {
                is FirebaseAuthError -> "FirebaseAuthError"
                is PlatformError -> "PlatformError"
                is PluginError -> "PluginError"
            }
        }
}

fun FirebaseAuthOAuthPluginError.toResult(result: Result) {
    when (this) {
        is FirebaseAuthOAuthPluginError.FirebaseAuthError ->
            result.error(code, exception.localizedMessage, exception)
        is FirebaseAuthOAuthPluginError.PlatformError ->
            result.error(code, exception.localizedMessage, exception)
        is FirebaseAuthOAuthPluginError.PluginError ->
            result.error(code, error, null)
    }
}

File: firebase_auth_oauth/ios/Classes/FirebaseAuthOAuthPluginError.swift
//
//  FirebaseAuthOAuthPluginError.swift
//  firebase_auth_oauth
//
//  Created by Amr Yousef on 20/06/2020.
//

enum FirebaseAuthOAuthPluginError: Error {
	case FirebaseAuthError(error: Error)
	case PluginError(error: String)
	case PlatformError(error: Error)
	
	var code: String {
		switch self {
		case .FirebaseAuthError:
			return "FirebaseAuthError"
		case .PluginError:
			return "PluginError"
		case .PlatformError:
			return "PlatformError"
		}
	}
}

extension FirebaseAuthOAuthPluginError {
	func flutterError() -> FlutterError {
		switch self {
		case .FirebaseAuthError(let error):
			return FlutterError(code: code, message: error.localizedDescription, details: nil)
		case .PlatformError(let error):
			return FlutterError(code: code, message: error.localizedDescription, details: nil)
		case .PluginError(let error):
			return FlutterError(code: code, message: error, details: nil)
		}
	}
}
